File: src/sender/email.py
#!/usr/bin/python
# Author: Diptesh Chatterjee
"""
Class defining the structure of an Email. This class is immutable and should
only be created using an EmailBuilder object. I know of no other way to make
classes immutable in Python.
"""
class Email:
    def __init__(self, sender="", to=[], cc=[], bcc=[], subject="", body="",
            attachments=[], mailFormat='text'):
        self.__sender__ = sender
        self.__to__ = to
        self.__cc__ = cc
        self.__bcc__ = bcc
        self.__subject__ = subject
        self.__body__ = body
        self.__attachments__ = attachments
        self.__format__ = mailFormat

    def getSender(self):
        return self.__sender__

    def getRecepients(self):
        return self.__to__

    def getCCList(self):
        return self.__cc__

    def getBCCList(self):
        return self.__bcc__

    def getSubject(self):
        return self.__subject__

    def getBody(self):
        return self.__body__

    def getAttachments(self):
        return self.__attachments__

    def getFormat(self):
        return self.__format__

"""
Class EmailBuilder implements a builder pattern for Email. Since Email is an
immutable class, I chose to use Builder pattern.
Usage:
    builder = EmailBuilder()
              .setSender('foo@example.com')
              .addReceiver('bar@anotherexample.com')
              .addReceiver('foobar@somedomain.com')
              .
              .
              .
              .create()
Mail format can only be text or HTML.
"""

class EmailBuilder:
    def __init__(self):
        self.__sender__ = ""
        self.__receivers__ = []
        self.__cc__ = []
        self.__bcc__ = []
        self.__subject__ = ""
        self.__body__ = ""
        self.__attachments__ = []
        self.__format__ = 'text'

    def create(self):
        return Email(self.__sender, self.__receivers__, self.__cc__,
                self.__bcc__, self.__subject__, self.__body__,
                self.__attachments__)

    def setSender(self, sender):
            self.__sender__ = sender

    def addReceiver(self, receiver):
            self.__receivers__.append(receiver)

    def addCC(self, cc):
            self.__cc__.append(cc)

    def addBCC(self, bcc):
            self.__bcc__.append(bcc)

    def addAttachment(self, attachment):
            self.__attachments__.append(attachment)

    def setSubject(self, subject):
            self.subject = subject

    def setBody(self, body):
            self.body = body

    def setFormat(self, mailFormat):
        if (not mailFormat.tolower() == 'text') and (not mailFormat.tolower()
                =='html'):
            raise Exception('Incorrect format. Can only be text or HTML')
        self.__format__ = mailFormat

File: src/main/java/com/diptesh/email/Email.java
package main.java.com.diptesh.email;

import java.util.ArrayList;
import java.util.List;

public final class Email {
	public static final class Builder {
		private final Email email;

		public Builder() {
			email = new Email();
		}

		public Builder addAttachment(final Attachment attach) {
			email.attachments.add(attach);
			return this;
		}

		public Builder addBcc(final Person bcc) {
			email.bccList.add(bcc);
			return this;
		}

		public Builder addCC(final Person cc) {
			email.ccList.add(cc);
			return this;
		}

		public Builder addRecepient(final Person recepient) {
			email.recepients.add(recepient);
			return this;
		}

		public Email build() {
			if (email.body == null || email.body.isEmpty()) {
				email.body = " ";
			}
			if (email.format == Format.HTML) {
				email.convertBodyToHtml();
			}
			return email;
		}

		public Builder setBody(final String body) {
			email.body = body;
			return this;
		}

		public Builder setFormat(final Format format) {
			email.format = format;
			return this;
		}

		public Builder setSender(final Person sender) {
			email.sender = sender;
			return this;
		}

		public Builder setSubject(final String subject) {
			email.subject = subject;
			return this;
		}
	};

	public static enum Format {
		HTML, TEXT
	}

	private final List<Attachment> attachments;
	private final List<Person> bccList;
	private String body;
	private final List<Person> ccList;
	private Format format;
	private final List<Person> recepients;
	private Person sender;
	private String subject;

	private Email() {
		sender = null;
		subject = null;
		body = null;
		format = null;
		recepients = new ArrayList<Person>();
		ccList = new ArrayList<Person>();
		bccList = new ArrayList<Person>();
		attachments = new ArrayList<Attachment>();
	}

	private void convertBodyToHtml() {
		final StringBuilder sb = new StringBuilder();
		sb.append("<html>");
		for (final String s : body.split("\n")) {
			sb.append(s);
			sb.append("<br>");
		}
		sb.append("</html>");
		body = sb.toString();
	}

	public List<Attachment> getAttachments() {
		return attachments;
	}

	public List<Person> getBccList() {
		return bccList;
	}

	public String getBody() {
		return body;
	}

	public List<Person> getCCList() {
		return ccList;
	}

	public Format getFormat() {
		return format;
	}

	public List<Person> getRecepients() {
		return recepients;
	}

	public Person getSender() {
		return sender;
	}

	public String getSubject() {
		return subject;
	}

	public boolean isValidEmail() {
		if (subject == null || subject.isEmpty()) {
			return false;
		}
		if (sender == null) {
			return false;
		}
		if (!sender.isValidPerson()) {
			return false;
		}
		for (final Person recepient : recepients) {
			if (!recepient.isValidPerson()) {
				return false;
			}
		}
		for (final Person cc : ccList) {
			if (!cc.isValidPerson()) {
				return false;
			}
		}
		for (final Person bcc : bccList) {
			if (!bcc.isValidPerson()) {
				return false;
			}
		}
		return true;
	}
}
